### ISSUE TITLE
Missing memory data

### ISSUE THREAD

**Issue by EvilPlankton** [autofitcloud/isitfit#10]
"optimize" is reporting quite a few "No memory data" servers but we have Datadog and the integration appears to be working. Debug output has lines like...

`Found datadog metrics in redis cache for i-...`

I have confirmed memory data appears in Datadog.

What do?

**Comment by shadiakiki1986** [comment]
If you use isitfit cost analyze --save-details, you get 2 csv files saved
locally with details (the program will output the filenames on the
console). Could you check if they contain a column for the memory data? The
column is called datadog.memory.used

**Comment by shadiakiki1986** [comment]
After some in-depth checks, I fixed a few issues, but I'm not 100% sure it would fix this for your case. Do you mind upgrading and checking again? Here are the commands to help you:
- `pip3 install --upgrade isitfit`: upgrades isitfit
- `isitfit version`: check that you have 0.18.11
- `isitfit cost optimize`: re-run this and check that the instances that have you verified to have datadog memory metrics no longer show "No memory data" in the table for recommended sizing

**Comment by EvilPlankton** [comment]
No luck. Still seeing a lot of "No memory data".

`Generated by isitfit version 0.18.11`

I am particularly interested in this integration and am happy to assist with debugging. What else can I do to help?

**Comment by shadiakiki1986** [comment]
The simplest thing to do would be if you can share detailed logs with me
over email. I can release a new version early next week with logs that will
be useful specifically for your case. Then you can upgrade and send me the
logs by email to shadi@autofitcloud.com and we take it from there. Sounds
good?

**Comment by EvilPlankton** [comment]
Sounds good. Will do.

**Comment by shadiakiki1986** [comment]
Hello. I just released isitfit version 0.19. This version contains a major revamp of how `isitfit` chooses to use datadog data or fall back on cloudwatch data if the datadog data was not available. It also provides more information to debug this issue.

Upon running `isitfit cost analyze` or `isitfit cost optimize`, you will see a table similar to the below *above* the results table.

```
Status of metric sources (for service ec2)
+-----------+--------------+------+
| datadog   | cloudwatch   |   ID |
|-----------+--------------+------|
| ok        | no need      |   10 |
| no data   | ok           |    5 |
| no data   | no data      |    2 |
+-----------+--------------+------+
Details at /tmp/isitfit/isitfit-sourceStatus-a547va3l.csv
```

This table will give us a general idea of how many EC2 instances are loading data from datadog and how many are falling back on Cloudwatch (and hence the `no memory data` note you were seeing in the table of `isitfit cost optimize`). Also, the detail csv file below this table will show each instance's ID along with the source being datadog or cloudwatch.

It would be great if you can run the below steps to help me debug this issue:
- `pip3 install --upgrade isitfit`: upgrades the local installation of isitfit
- `isitfit vesion`: should show version 0.19
- `isitfit cost optimize`: re-run the `optimize` command and check if the problem persists with this new version
- If the problem is fixed, then party :tada:
- Otherwise, re-run the command as follows (replacing `profilename` with your own profile)
```
isitfit --debug cost --profile=profilename --ndays=7 optimize 2>/tmp/isitfit/issue_10-logs.txt
```
- The above would save the logs to a local file. It would be great if you can send me (on this issue or to my email shadi@autofitcloud.com) the output of the following command run on this log dump (replacing `example_instance_id` with an example instance ID that you see with `No ram` and you know to have memory metrics in your datadog dashboard)

```
grep Datadog /tmp/isitfit/issue_10-logs.txt | grep example_instance_id | head
```
For example, it would output
```
Datadog env vars available
Datadog: data not found for i-e1ca46eb: No cpu_idle_min found for i-e1ca46eb
```

**Comment by EvilPlankton** [comment]
Got the same output as in your example:
`# grep Datadog isitfit.log
Datadog env vars available
Datadog: data not found for i-088f4f4eedda4b69d: No cpu_idle_min found for i-088f4f4eedda4b69d`
...

In Datadog I see metric data for system.cpu.idle, but no "cpu_idle_min".

**Comment by shadiakiki1986** [comment]
Thanks for the output. I found a few other issues and fixed them in version `0.19.3` released just now. Do you mind upgrading isitfit and following the same steps listed above again?

**Comment by shadiakiki1986** [comment]
Hello @EvilPlankton  and happy new year! Did you have time to check if this issue is solved for you?

**Comment by EvilPlankton** [comment]
Sent you an email with latest results.

**Comment by shadiakiki1986** [comment]
This issue is now resolved as of isitfit 0.19.19.

It turns out that the datadog hostname is not necessarily the same as the AWS ID,
so isitfit needed an extra mapping step between the two.

Relevant docs

- https://docs.datadoghq.com/api/?lang=bash#search-hosts
- https://docs.datadoghq.com/agent/faq/how-datadog-agent-determines-the-hostname/?tab=agentv6v7#potential-host-names

To commemorate the fierce battle against this issue, I dedicate this haiku:

```
Mystery deepened
Unbreakable yet broken
Make IDs map again
```
